### ISSUE TITLE
Running in Tomcat

### ISSUE THREAD

**Issue by punkfairy** [medusa-project/cantaloupe#195]
Is it still possible to deploy the WAR with Tomcat? It seems like there are quite a few dependencies on Jetty, and it seems like there are problems shutting down when deploying in Tomcat (there are lots of Threads left running and Tomcat can't shut it down.)

**Comment by adolski** [comment]
Thanks for the heads-up, @punkfairy . I will look into this.

The Jetty server is used in standalone mode only. The client is also used by HttpResolver and the client and server share some other Jetty dependencies.

**Comment by punkfairy** [comment]
Would the jstack output post-shutdown attempt be helpful?

**Comment by punkfairy** [comment]
It looks like all of the problem non-daemon Threads trace back to HttpClient in Jetty:

```
"HttpClient@1057328382-115" #115 prio=5 os_prio=31 tid=0x00007fc5d8b29800 nid=0x13303 waiting on condition [0x0000000134ee8000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000744c87a90> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.reservedWait(ReservedThreadExecutor.java:193)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:225)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-114" #114 prio=5 os_prio=31 tid=0x00007fc5d8c40000 nid=0x13103 waiting on condition [0x0000000134de5000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007449bdf60> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:392)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:571)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$800(QueuedThreadPool.java:50)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:634)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-113" #113 prio=5 os_prio=31 tid=0x00007fc5d88d1800 nid=0x12f03 waiting on condition [0x0000000134ce2000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007449be1c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.reservedWait(ReservedThreadExecutor.java:193)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:225)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-111" #111 prio=5 os_prio=31 tid=0x00007fc5d8a24000 nid=0xaa0f runnable [0x0000000134adc000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
	at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
	at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:117)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000007449be3f0> (a sun.nio.ch.Util$3)
	- locked <0x00000007449be400> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000007449be3a0> (a sun.nio.ch.KQueueSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select(ManagedSelector.java:346)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:299)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:179)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.io.ManagedSelector$$Lambda$82/1171854681.run(Unknown Source)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-110" #110 prio=5 os_prio=31 tid=0x00007fc5d345e800 nid=0xe707 runnable [0x0000000133bad000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
	at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
	at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:117)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x0000000744a28250> (a sun.nio.ch.Util$3)
	- locked <0x0000000744a28260> (a java.util.Collections$UnmodifiableSet)
	- locked <0x0000000744a28200> (a sun.nio.ch.KQueueSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select(ManagedSelector.java:346)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:299)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:179)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.io.ManagedSelector$$Lambda$82/1171854681.run(Unknown Source)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-109" #109 prio=5 os_prio=31 tid=0x00007fc5d3465800 nid=0xa00f runnable [0x00000001333a1000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.KQueueArrayWrapper.kevent0(Native Method)
	at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
	at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:117)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x0000000744a49bb0> (a sun.nio.ch.Util$3)
	- locked <0x0000000744a49bc0> (a java.util.Collections$UnmodifiableSet)
	- locked <0x0000000744a49b60> (a sun.nio.ch.KQueueSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select(ManagedSelector.java:346)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:299)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:179)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.io.ManagedSelector$$Lambda$82/1171854681.run(Unknown Source)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)

"HttpClient@1057328382-108" #108 prio=5 os_prio=31 tid=0x00007fc5d36ef800 nid=0xdd0b waiting on condition [0x000000013309e000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007449bdf60> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:392)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:571)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$800(QueuedThreadPool.java:50)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:634)
	at java.lang.Thread.run(Thread.java:748)

```

Oh, also this is the latest 3.4 release (3.4.2 I think?)

**Comment by adolski** [comment]
That makes sense. I'll bet the problem is that `stop()` is never called on the Jetty client in HttpResolver. Needs to be done in a context listener.

**Comment by punkfairy** [comment]
We tried a quick test where we dropped this into `ApplicationContextListener.contextDestroyed()`:
```
        try {
            HttpResolver.doStop();
        } catch (Exception e) {
            System.err.println("Trying to kill HttpClient: " + e.getMessage());
        }
```
with this in HttpResolver:
```
public class HttpResolver extends AbstractResolver implements StreamResolver {

    public static synchronized void doStop() throws Exception {
        jettyClient.stop();
    }
```
and it definitely fixed the problem.

**Comment by adolski** [comment]
Thanks @punkfairy ! My own findings concur. This should be fixed now, in `release/3.4`. Thanks again  for reporting.

**Comment by punkfairy** [comment]
Works! Thank you!
